packages/contracts-bedrock/src/interop/L2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { L2CrossDomainMessenger as LegacyL2CrossDomainMessenger } from "src/L2/L2CrossDomainMessenger.sol";
+import { Constants } from "src/libraries/Constants.sol";
+import { Encoding } from "src/libraries/Encoding.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { SafeCall } from "src/libraries/SafeCall.sol";
+
+/// @title L2CrossDomainMessenger
+/// @notice L2CrossDomainMessenger is an extended version of the existing predeploy that supports
+///         interopability between many chains (L2-L2). Since we are keeping the existing messaging
+///         contracts the same while iterating on the L2-L2 implementation, changes to the interface
+///         and messaging format that would reside in the CrossDomainMessenger are scoped internally
+///         in this contract. The L2-L1 flow remains unchanged.
+contract L2CrossDomainMessenger {
+    // CrossDomainMessenger: some copied internals & updated dispatch/message spec
+
+    /// @notice Latest message version identifier (interop-enabled)
+    uint16 public constant MESSAGE_VERSION = 2;
+
+    /// @notice Chain identifier of the network
+    bytes32 public immutable CHAIN_ID;
+
+    /// @param _chain_id Chain identifier of this network
+    constructor(bytes32 _chain_id) {
+        CHAIN_ID = _chain_id;
+    }
+
+    // TODO: Events
+
+    /// @notice Nonce for the next message to be sent
+    uint240 internal msgNonce;
+
+    /// @notice Address of the sender of the currently executing message
+    address internal xDomainMsgSender;
+
+    /// @notice Mapping of succesfully delivered messages
+    mapping(bytes32 => bool) public successfulMessages;
+
+    /// @notice Mapping of delivered messages in a failed state
+    mapping(bytes32 => bool) public failedMessages;
+
+    /// @notice Sends a message to the specified destination chain. If the destination chain is ETH Mainnet (0x1), then
+    ///         this message follows the legacy message flow and formates by forwarding to the existing
+    ///         L2CrossDomainMessenger predeploy. When this implementation is ready as a replacement, L2ToL2 and L2ToL1
+    ///         is natively handled by the respective internal message passer contracts.
+    function sendMessage(
+        bytes32 _destination,
+        address _target,
+        bytes calldata _message,
+        uint32 _minGasLimit
+    )
+        external
+        payable
+    {
+        // L2->L1 Support: Required as we are leaving the CrossDomainMessenger untouched; All L2->L1 messages will still
+        // be using the old message format (Version 1), hence forwarding to the "legacy" contract.
+        bytes32 ETH_MAINNET_ID = bytes32(hex"01");
```

Comment:
Do you know if the 1 here is left aligned or right aligned? This kind of thing happens in solidity depending on the types that you use. I recommend `bytes32(uint256(1))` so that its right aligned

Reply:
yea good call out, unsure & would have to verify alignment